### ISSUE TITLE
Rails 7.2 Issue

### ISSUE THREAD

**Issue by denisahearn** [zdennis/activerecord-import#847]
I have a Rails app that uses the `activerecord-import`, `state_machines` and `state_machines-activerecord` gems. I tried upgrading this app Rails `7.0.8` to Rails `7.2.1`, and noticed a difference in behavior with the `bulk_import` method when updating existing records.

With Rails `7.0.8`, when updating a record via `bulk_import`, if the record gets updated in a callback via a state machine event, the updates attributes are saved to the database by `bulk_import`. With Rails `7.2.1`, the changes to the record made by the state machine event are not getting saved to the database by `bulk_import`.

I created a simple Rails app to demonstrate the issue. Attached are two .zip files containing the app, each with a specific version of Rails:

* `activerecord-import-issue-rails-7.0.8.zip` - Uses Rails `7.0.8` and demonstrates the expected behavior of `bulk_import`

```
➜  activerecord-import-issue git:(main) rails --version
Rails 7.0.8.4
➜  activerecord-import-issue git:(main) bin/rails test test/models/order_test.rb
Running 2 tests in a single process (parallelization threshold is 50)
Run options: --seed 40205

# Running:

..

Finished in 0.060363s, 33.1329 runs/s, 82.8322 assertions/s.
2 runs, 5 assertions, 0 failures, 0 errors, 0 skips
➜
```

* `activerecord-import-issue-rails-7.2.zip` - Uses Rails `7.2.1` and demonstrates unexpected behavior of `bulk_import`

```
➜  activerecord-import-issue git:(rails-7.2) rails --version
Rails 7.2.1
➜  activerecord-import-issue git:(rails-7.2) bin/rails test test/models/order_test.rb
Running 2 tests in a single process (parallelization threshold is 50)
Run options: --seed 28721

# Running:

F

Failure:
OrderTest#test_upsert_object_status_event_with_bulk_import [test/models/order_test.rb:31]:
Expected false to be truthy.


bin/rails test test/models/order_test.rb:15

.

Finished in 0.030592s, 65.3766 runs/s, 163.4414 assertions/s.
2 runs, 5 assertions, 1 failures, 0 errors, 0 skips
```

The demo app has an `Order` model with a state machine defined in it:

```
class Order < ApplicationRecord
  state_machine :status, initial: :pending do
    event :upsert_object do
      transition :pending => :completed, if: ->(order) { order.completable? }
      transition :pending => same
    end
  end

  def completable?
    completed_at.present?
  end
end
```

and has these two tests:

```
require "test_helper"

class OrderTest < ActiveSupport::TestCase
  test "upsert_object status event with ActiveRecord save" do
    order = Order.create(product: "My Product")
    assert order.status == "pending"

    order.status_event = "upsert_object"
    order.completed_at = DateTime.now
    order.save

    assert order.status == "completed"
  end

  test "upsert_object status event with bulk import" do
    order = Order.new(product: "My Product")
    assert order.status == "pending"

    order.status_event = "upsert_object"
    order.completed_at = DateTime.now
    result = Order.bulk_import(
      [order],
      batch_size: 1,
      validate: true,
      on_duplicate_key_update: %i[product status completed_at updated_at],
      validate_with_context: :mass_update,
      timestamps: true
    )

    assert result.failed_instances.none?
    assert order.status == "completed"
  end
end
```

With Rails `7.0.8`, both tests pass. With Rails `7.2.1`, the `upsert_object status event with ActiveRecord save` test still passes, however the `upsert_object status event with bulk import` test now fails on the `assert order.status == "completed"` assertion.

Debugging the issue, I found that if I make a small change to `lib/activerecord-import/import.rb` I can fix the failing test with Rails `7.2.1`. If I add a new line after https://github.com/zdennis/activerecord-import/blob/master/lib/activerecord-import/import.rb#L110 such that the block returns `true` then the test passes.

So, change this:

```
else # ActiveRecord 5.1
  # Note that this is a gross simplification of ActiveSupport::Callbacks#run_callbacks.
  # It's technically possible for there to exist an "around" callback in the
  # :validate chain, but this would be an aberration, since Rails doesn't define
  # "around_validate". Still, rather than silently ignoring such callbacks, we
  # explicitly raise a RuntimeError, since activerecord-import was asked to perform
  # validations and it's unable to do so.
  #
  # The alternative here would be to copy-and-paste the bulk of the
  # ActiveSupport::Callbacks#run_callbacks method, which is undesirable if there's
  # no real-world use case for it.
  raise "The :validate callback chain contains an 'around' callback, which is unsupported" unless runner.final?
  runner.invoke_before(env)
  runner.invoke_after(env)
```

to this:

```
else # ActiveRecord 5.1
  # Note that this is a gross simplification of ActiveSupport::Callbacks#run_callbacks.
  # It's technically possible for there to exist an "around" callback in the
  # :validate chain, but this would be an aberration, since Rails doesn't define
  # "around_validate". Still, rather than silently ignoring such callbacks, we
  # explicitly raise a RuntimeError, since activerecord-import was asked to perform
  # validations and it's unable to do so.
  #
  # The alternative here would be to copy-and-paste the bulk of the
  # ActiveSupport::Callbacks#run_callbacks method, which is undesirable if there's
  # no real-world use case for it.
  raise "The :validate callback chain contains an 'around' callback, which is unsupported" unless runner.final?
  runner.invoke_before(env)
  runner.invoke_after(env)
  true # <== THIS FIXES MY FAILING TEST
```

I decided to open an issue instead of a PR because I honestly am not sure if this is the correct fix to make. I'm hoping that by providing enough details someone with more knowledge of the inner workings of `activerecord-import` will be able to understand the problem and make the correct fix if it's not this.

[activerecord-import-issue-rails-7.0.8.zip](https://github.com/user-attachments/files/16784811/activerecord-import-issue-rails-7.0.8.zip)

[activerecord-import-issue-rails-7.2.zip](https://github.com/user-attachments/files/16784819/activerecord-import-issue-rails-7.2.zip)

**Comment by jkowens** [comment]
@denisahearn can you confirm you are using activerecord-import 1.8.0?

**Comment by denisahearn** [comment]
@jkowens Yes, my Rails app, along with the sample Rails app I created to demonstrate the issue are both using `activerecord-import 1.8.0`

**Comment by jkowens** [comment]
@denisahearn thank you, the sample apps will be very helpful. Rails 7.2 made significant changes to validation callbacks, so we'll have to step through that. Have you noticed specific validations that are preventing the record to save?

**Comment by denisahearn** [comment]
@jkowens No, no validations were preventing the record to save. You'll see when you look at the sample Rails app I attached to this issue (and shown above in the issue description) that the simple `Order` model used to demonstrate the problem has no validations defined in it. 

Note that this model does have a state machine defined in it though (via the `state_machines-activerecord` gem) that gets run during the callbacks, that updates the value of a `status` attribute on the order record being updated by `bulk_import`. As far as I could tell, nothing in the state machine transition was causing the callback to fail.

**Comment by jkowens** [comment]
@denisahearn right, I haven't used the state_machines-activerecord gem in quite some time. I assume it is adding validation callbacks behind the scenes correct? I'm not entirely sure, but my guess is that returning `true` from the block essentially bypasses validation.

**Comment by denisahearn** [comment]
@jkowens I'm not very familiar with the internal workings of the `state_machines-activerecord` gem, but https://github.com/state-machines/state_machines-activerecord/blob/master/lib/state_machines/integrations/active_record.rb#L51 indicates you are may be correct.

I guess it's possible that the fault lies in the `state_machines-activerecord` gem and not with `activerecord-import` regarding my specific issue. Please let me know if you determine that `activerecord-import` is functioning properly with the callback changes in Rails 7.2, and I will see if I can pin the cause on the other gem.

**Comment by denisahearn** [comment]
@jkowens FWIW, this is the line of code in the `state_machines` gem (which `state_machines-activerecord` sits on top of) that made me think the callback block in `activerecord-import` needed to return `true`:

https://github.com/state-machines/state_machines/blob/master/lib/state_machines/transition_collection.rb#L153

**Comment by jkowens** [comment]
Ah interesting, it looks like `invoke_after` used to always return an `Array` now it can return `nil` so before the result would always evaluate true.

**Comment by jkowens** [comment]
@denisahearn your fix is probably fine, for the sake of backwards compatibility I think I'm going to go with this:

```
  raise "The :validate callback chain contains an 'around' callback, which is unsupported" unless runner.final?
  runner.invoke_before(env)
  runner.invoke_after(env) || []
  ```

**Comment by jkowens** [comment]
@denisahearn I just pushed an update to master that should resolve this issue. I'll cut a release this week.

**Comment by denisahearn** [comment]
@jkowens Thanks for promptly addressing this issue. I'll grab the new release after you publish it.

**Comment by jkowens** [comment]
Sorry for the delay. Fix released in v1.8.1 👍
